### ISSUE TITLE
Vertical Page Fitting

### ISSUE THREAD

**Issue by jackcrymble** [dc-js/dc.datatables.js#3]
Suggested Enhancement: Vertical Page Fitting

In datatables you can set the number of records shown per page with the top left drop-down menu. This feature is useful if you adjust the size of the container based on number of rows shown. In a case where the container has a set size and you want the datatable to fill the vertical space I believe it would be of use to have an option to set the table to fill vertically. 

I've come across this plug in on datatables.net which fit's the datatable to the vertical size of its container:
https://datatables.net/blog/2015-04-10

Could this be added to the dc.datatables package? I presume the option could be set through the exposed dt object?

Thanks

Jack

**Comment by gordonwoodhull** [comment]
Very cool. This will be a good test case of using datatables plugins and setting datatables options.

The library should be designed so that plugins and special options can be used without modifying the library. I don't know if that is true yet. 

I think you should be able to load the plugin just by making sure the linked script runs before you initialize the table. 

It seems that options can only be set when initializing the data table, [according to the manual](https://datatables.net/manual/options).

Would you be willing to try adding the option directly to the source, in the `render` function above `columns:`? If loading the plugin yourself and adding the option works, we can add an `options` getter/setter which adds to the options sent when DataTables is initialized.

**Comment by jackcrymble** [comment]
Hi @gordonwoodhull,

In my Angular 6 project I've installed jquery and added the plugin to the index.html file.

`  <script type="text/javascript" src="https://cdn.datatables.net/plug-ins/1.10.19/features/pageResize/dataTables.pageResize.min.js"></script>`

This is trying to load the plugin but an error occurs at run time:
![image](https://user-images.githubusercontent.com/11898127/46792095-6de01780-cd3a-11e8-8d87-46eca83d4cb1.png)
Referring to this line of code:
![image](https://user-images.githubusercontent.com/11898127/46792145-8fd99a00-cd3a-11e8-8142-f781ddf027e2.png)

Perhaps this plug in is not yet compatible with Angular 6?

**Comment by jackcrymble** [comment]
I have, however, found that if you add options directly to the dc.datatables.js file (in render() before columns: ) then they are picked up by the datatable and applied.

```
_dt = jQuery(table.node()).DataTable({
            "scrollY": "200px",
            "scrollCollapse": true,
            "paging": false,
            columns: _table.columns().map(function(c) {
                var col = {
```
So the ability to add options would be possible with built-in [datatables options](https://datatables.net/reference/option/).

**Comment by jackcrymble** [comment]
I propose three changes which work together to allow project side selection of options:

1. Adding an `_options` `var` to the dc_datatables.datatable object

```
var _table = {}, // this object
        _dt, // jquery.dataTables object
        _root, // selected div
        _dimension, // crossfilter dimension
        _options, // options for datatable customisation
        _group, _size, _columns, _sortBy, _order; // for compatibility; currently unused
```

2. Add function to allow project side to specify `_options`
```
_table.options = function(_) { 
      if(!arguments.length) {
          return _options;
      }
      _options = _;
      return this;
  };
```

3. Adjust `_table.render()` to merge `columns:` with the `_options` provided:
```
_dt = jQuery(table.node()).DataTable({
            columns: _table.columns().map(function(c) {
                var col = {
                    name: typeof c === 'string' ? c : c.label,
                    type: typeof c === 'object' ? c.type : 'num',
                    render: columnRenderer(c)
                };
                col.title = col.name.charAt(0).toUpperCase() + col.name.slice(1);
                return col;
            }),
            ..._options
```

What do you think?

**Comment by gordonwoodhull** [comment]
These changes make a lot of sense.

However, I take it you were not able to get the plugin to work? It looks like the plugin uses exactly the same mechanism to modify DataTables that DataTables uses to modify jQuery - so I think there should be a way to make this work.

Did you add the plugin in the same way that you added dc.datatables.js and/or jQuery?

Are answers like this one helpful? https://stackoverflow.com/questions/16935095/correct-way-to-integrate-jquery-plugins-in-angularjs

I think you need to tie in to Angular's dependencies/modules somehow, rather than just including the script.

**Comment by jackcrymble** [comment]
That is correct, I could not get the plugin working within Angular 6. 

- I have tried adding a link to the plugin inside the angular.json as part of the scripts array. This cases angular to look for the file locally.
- Using a <script> tag to add the js file within the index.html. This is the example i gave in https://github.com/dc-js/dc.datatables.js/issues/3#issuecomment-428882737 
- adding the jquery CDN inside a <script> tag before the pageResize CDN. This caused a PageResize undefined error.

Unfortunately, I'm coming short on ways to further this for plugins...

**Comment by gordonwoodhull** [comment]
The world seems to think that AMD (the format of this plugin) is directly compatible with webpack (what Angular uses).

I was able to get it to *build* by copying the plugin to `src/app` and changing it slightly. I don't know if it will run, because I don't know enough Angular yet.

The plugin tries to import datatables like this:

```js
		// AMD
		define( ['jquery', 'datatables.net'], function ( $ ) {
			return factory( $, window, document );
		} );
```

That's an invalid module name for ES6, and it's inconsistent with how the rest of our project load datatables.

Changing it like so:

```js
		// AMD
		define( ['jquery', 'datatables'], function ( $ ) {
			return factory( $, window, document );
		} );
```

and testing it like this:

```ts
import * as pageResize from './dataTables.pageResize.js';
// ...
console.log('pr', pageResize);
```

It does build: https://github.com/dc-js/ang-dctable

The object it prints is empty but that is to be expected because this plugin changes jQuery.

I bet that webpack has some way to map module names, so that they don't have to be modified.

**Comment by jackcrymble** [comment]
Hi @gordonwoodhull 

I have been playing around with this and have been able to get it working following your steps in the app side of my project with aid of the `options` in my pull request.

I had previously been trying to get this working from an in house library which is pulled in by each project being worked on. I'm unsure why the import does not work from the library side, but I don't believe this is a datatables problem.

With the additional options code, I am able to successfully import and pass across the pageResize and lengthchange options like so :
```
options: {pageResize: true,
                lengthChange: false}
```

**Comment by gordonwoodhull** [comment]
Very cool, glad you got it working. I merged PR #4 for version 0.0.7, so I think this issue is closed.

Thanks for your contribution and feedback!
